src/gui/widgets/LinkedModelGroupLayout.cpp

```diff
@@ -0,0 +1,303 @@
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutVerticalSpacing);
+	}
+}
+
+int LinkedModelGroupLayout::count() const
+{
+	return m_itemMap.size() - 1;
+}
+
+QMap<QString, QLayoutItem*>::const_iterator
+LinkedModelGroupLayout::pairAt(int index) const
+{
+	if (index < 0)
+		return m_itemMap.cend();
+	QMap<QString, QLayoutItem*>::const_iterator itr = m_itemMap.cbegin();
+	++itr; // skip search bar
+	while (index-->0 && itr != m_itemMap.cend()) { ++itr; }
+	return itr;
+}
+
+// linear time :-(
+QLayoutItem *LinkedModelGroupLayout::itemAt(int index) const
+{
+	auto itr = pairAt(index);
+	return itr == m_itemMap.end() ? nullptr : itr.value();
+}
+
+QLayoutItem *LinkedModelGroupLayout::itemByString(const QString &key) const
+{
+	auto itr = m_itemMap.find(key);
+	return (itr == m_itemMap.end()) ? nullptr : *itr;
+}
+
+// linear time :-(
+QLayoutItem *LinkedModelGroupLayout::takeAt(int index)
+{
+	auto itr = pairAt(index);
+	return (itr == m_itemMap.end()) ? nullptr : m_itemMap.take(itr.key());
+}
+
+Qt::Orientations LinkedModelGroupLayout::expandingDirections() const
+{
+	return Qt::Orientations();
+}
+
+bool LinkedModelGroupLayout::hasHeightForWidth() const
+{
+	return true;
+}
+
+int LinkedModelGroupLayout::heightForWidth(int width) const
+{
+	int height = doLayout(QRect(0, 0, width, 0), true);
+	return height;
+}
+
+void LinkedModelGroupLayout::setGeometry(const QRect &rect)
+{
+	QLayout::setGeometry(rect);
+	doLayout(rect, false);
+}
+
+QSize LinkedModelGroupLayout::sizeHint() const
+{
+	return minimumSize();
+}
+
+QSize LinkedModelGroupLayout::minimumSize() const
+{
+	// original formula from Qt's FlowLayout example:
+	// get maximum height and width for all children.
+	// as Qt will later call heightForWidth, only the width here really matters
+	QSize size;
+	for (const QLayoutItem *item : qAsConst(m_itemMap))
+	{
+		size = size.expandedTo(item->minimumSize());
+	}
+	const QMargins margins = contentsMargins();
+	size += QSize(margins.left() + margins.right(), margins.top() + margins.bottom());
+
+	// the original formula would leed to ~1 widget per row
+	// bash it at least to 400 so we have ~4 knobs per row
+	size.setWidth(qMax(size.width(), m_minWidth));
+	return size;
+}
+
+int LinkedModelGroupLayout::doLayout(const QRect &rect, bool testOnly) const
+{
+	int left, top, right, bottom;
+	getContentsMargins(&left, &top, &right, &bottom);
+	QRect effectiveRect = rect.adjusted(+left, +top, -right, -bottom);
+	int x = effectiveRect.x();
+	int y = effectiveRect.y();
+	int lineHeight = 0;
+
+	const QString filterText = m_searchBar->text();
+	bool first = true;
+
+	QMapIterator<QString, QLayoutItem*> itr(m_itemMap);
+	while (itr.hasNext())
+	{
+		itr.next();
+		QLayoutItem* item = itr.value();
+		QWidget *wid = item->widget();
+
+		if (first || // do not filter search bar
+			filterText.isEmpty() || // no filter - pass all
+			itr.key().contains(filterText, Qt::CaseInsensitive))
+		{
+			if (wid)
+			{
+				if (first)
```

Comment:
This could be combined with the above if since there is no else there.

Reply:
It can be even wrapped around the above if (and re-used for the else).